### ISSUE TITLE
potential bug in setting `positions` value for new token during decoding

### ISSUE THREAD

**Issue by thuann2cats** [Physical-Intelligence/openpi#705]
Hi OpenPi team,

I don’t quite understand this line in `pi0_fast.py` that’s intended to set the `positions` value for a new token to be passed in the next `.llm(…)` call during decoding, specifically why `+ step + 1`was needed instead of just `+ step`. https://github.com/Physical-Intelligence/openpi/blob/0d3eb2db836e443a6ca9debb242d0aabe4628f1f/src/openpi/models/pi0_fast.py#L293

I think this might be a bug. Let me walk through my reasoning, which begins on the `left_to_right_align()` call, where we roll the token embeddings, prefix mask and prefix attention mask to be right-aligned.

For the sake of example, let’s say
``` 
# Example: batch_size=2, total sequence length=8
# After right-alignment:
prefix_mask = jnp.array([
    [0, 0, 0, 0, 0, 1, 1, 1],  # Batch 0: 3 valid tokens at positions 5,6,7
    [0, 0, 0, 1, 1, 1, 1, 1]   # Batch 1: 5 valid tokens at positions 3,4,5,6,7
])

prefill_size = 8                    # Total sequence length
prefill_len = jnp.sum(prefix_mask, axis=-1)  # [3, 5] - number of valid tokens
prefix_start = prefill_size - prefill_len     # [5, 3] - start position of valid tokens
```
Then `prefix_positions` will have the following values:
```
prefix_positions = jnp.cumsum(prefix_mask, axis=-1) - 1

# Step by step:
cumsum_result = jnp.cumsum(prefix_mask, axis=-1)
# cumsum_result = [
#     [0, 0, 0, 0, 0, 1, 2, 3],  # Cumulative count of valid tokens
#     [0, 0, 0, 1, 2, 3, 4, 5]   # Cumulative count of valid tokens
# ]  --> this gives us the 1-indexed counts

prefix_positions = cumsum_result - 1
# prefix_positions = [
#     [-1, -1, -1, -1, -1, 0, 1, 2],  # Position indices for valid tokens
#     [-1, -1, -1, 0, 1, 2, 3, 4]     # Position indices for valid tokens  
# ] --> this gives us the 0-indexed positions for positional encoding
```
During the first generation step – the first new token (where `step = 0`):
```
# During generation step = 0 (first new token)
positions = prefill_len[:, None] + step + 1
# prefill_len = [3, 5], step = 0
# positions = [3, 5] + 0 + 1 = [4, 6]
```
So that means for first sequence in the batch:
```
# Positions 5,6,7 in sequence have RoPE positions: 0, 1, 2

# Step 0: position = 3 + 0 + 1 = 4 → But this should be 3
# This creates a gap: 0, 1, 2, and then 4
```
Similarly, for the second sequence in the batch:
```
# Positions 3,4,5,6,7 in sequence have RoPE positions: 0, 1, 2, 3, 4

# Step 0: position = 5 + 0 + 1 = 6 → But this should be 5  
# This creates a gap: 0, 1, 2, 3, 4 → 6
```
That’s why I think that line should be ` positions = prefill_len[:, None] + step ` instead. I was wondering if my understanding is correct, and if there was a reason for the `+ 1` in line 284.

Thanks!

**Comment by BrunoFANG1** [comment]
You are right on this, I also noticed this bug.
